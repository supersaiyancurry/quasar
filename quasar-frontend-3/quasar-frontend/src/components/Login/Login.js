/* eslint-disable no-nested-ternary */
import React, { useState } from 'react';
import { useHistory, Link } from 'react-router-dom';
import LogRocket from 'logrocket';
import { useAppContext } from '../utils/Context';
import { QUASAR_API, ERROR } from '../Constants';
// eslint-disable-next-line import/no-unresolved
import styles from './Login.module.css';
import FormField from '../form field/FormField';
import ErrorPage from '../error page/ErrorPage';
import Button from '../button/Button';
/**
 * Component for handling user login attempts.
 */
const Login = () => {
  const [password, setPassword] = useState('');
  const [email, setEmail] = useState('');
  const [error, setError] = useState(false);
  const [errorCode, setErrorCode] = useState(false);
  const history = useHistory();
  const { setIsLoggedIn, setUserEmail } = useAppContext();

  /**
     * Posts email and password to db to see if the information is valid.
     * @param {e} mouse click event
     */
  const handleSubmit = async (event) => {
    setError(false);
    event.preventDefault();
    const customerLogin = {
      email,
      password
    };
    let serverError;
    await fetch(`${QUASAR_API}/customers/login`, {
      method: 'POST',
      headers: new Headers({
        'Content-Type': 'application/json'
      }),
      body: (JSON.stringify(customerLogin))
    })
      .then(async (response) => {
        if (!response.ok) {
          serverError = response.status;
          throw new Error(<p>{ERROR}</p>);
        }
        const token = await response.text();
        sessionStorage.setItem('token', token);
        // parse JWT for payload containing user email
        const user = JSON.parse(atob(token.split('.')[1]));
        // set user email in storage
        setIsLoggedIn(true);
        setUserEmail(user.sub);
        history.push('/');
        LogRocket.log(`User logged in with the email '${user.sub}'`);
      })
      .catch((credentials) => {
        if (credentials !== customerLogin || customerLogin === '') {
          LogRocket.error('Invalid email or password.');
          setError(true);
        }
        // if server doesn't respond with an error message, display a 503
        setErrorCode(serverError || 503);
        LogRocket.error(serverError || '503 error: the service is unavailable.');
      });
  };
  return (
    <>
      {/* nested ternary operator that checks for an error code, if there is no error code
      the title will be displayed, if there is an error code and the error code is a
      404 it will display the title, if it is not a 404, it will display the error code with
      an error message */}
      {!errorCode
        ? (<div className={`${styles.login}`}><p>Welcome to the login page!</p></div>)
        : errorCode && (errorCode === 404 || errorCode === 400)
          ? (<div className={`${styles.login}`}>Welcome to the login page!</div>)
          : errorCode && errorCode !== 404
            ? (<div className={`${styles.code}`}><ErrorPage errorCode={errorCode} /></div>)
            : (<div className={`${styles.login}`}>Welcome to the login page!</div>)}
      <div>
        <div className={`jumbotron ${styles.jumbotron}`}>
          <form onSubmit={handleSubmit}>
            <h2 className={styles.title}>Login</h2>
            <div>
              {/* an error message is diaplyed if the login details to not match a valid user */}
              {(errorCode === 404 || errorCode === 400) && error
                ? <h5 className={styles.error}> Invalid email or password</h5>
                : null}
            </div>
            <FormField
              type="email"
              name="Email"
              value={email}
              changeCallback={setEmail}
            />
            <FormField
              type="password"
              name="Password"
              value={password}
              changeCallback={setPassword}
            />
            <Button type="submit" text="Submit" />
            <div className="mt-3">
              New customer?
              <Link id="register" to="/signup" className={`${styles.link}`}>Sign up here!</Link>
            </div>
          </form>
        </div>
      </div>
    </>
  );
};
export default Login;
