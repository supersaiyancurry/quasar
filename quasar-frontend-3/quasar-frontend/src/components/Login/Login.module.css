body {
    background-color: #F3E0DC;
}

.jumbotron {
    background-color: #D4A59A;
    margin: auto;
    margin-top: 95px;
    border-radius: 10px;
    max-width: 700px;
}

.title {
    margin-top: -40px;
    margin-bottom: 40px;
    font-size: 35px;
    text-align: center;
    color: black;
}

.link {
    color: #5C2018;
    padding-left: 0.3rem;
    font-weight: 500;
}

.error {
    color: red;
    margin-top: -35px;
    margin-left: 12.4rem;
    margin-bottom: 10px;
    position: absolute;
}

.code {
    margin-top: -1.5%;
    position: absolute;
    justify-content: center;
    width: 100%;
}

.login {
    margin-top: .8%;
    position: absolute;
    justify-content: center;
    width: 100%;
    font-size: 2em;
    font-weight: 500;
    color: #4285F4;
}

.loading {
    font-size: 30px;
    font-weight: bold;
    position: relative
}