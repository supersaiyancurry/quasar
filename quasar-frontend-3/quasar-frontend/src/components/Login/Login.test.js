import React from 'react';
import { render, screen, cleanup } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
// eslint-disable-next-line import/no-unresolved
import Login from './Login';

afterEach(cleanup);

const renderWithRouter = (component) => {
  const history = createMemoryHistory();
  return {
    ...render(
      <Router history={history}>
        {component}
      </Router>
    )
  };
};

it('renders without crashing', () => {
  renderWithRouter(
    <Login />
  );
  expect(screen.getByRole('button', { name: /Login/i }));
});

it('header component matches snapshot', () => {
  const { asFragment } = renderWithRouter(
    <Login />
  );

  expect(asFragment()).toMatchSnapshot();
});

it('has the correct URL', () => {
  global.window = { location: { pathname: null } };
  expect(global.window.location.pathname).toEqual('/');
});
