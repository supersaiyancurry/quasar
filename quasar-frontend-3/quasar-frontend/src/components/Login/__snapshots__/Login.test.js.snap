// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="login"
  >
    <p>
      Welcome to the login page!
    </p>
  </div>
  <div>
    <div
      class="jumbotron jumbotron"
    >
      <form>
        <h2
          class="title"
        >
          Login
        </h2>
        <div />
        <div
          class="form-group row"
        >
          <label
            class="col-sm-3 col-form-label label"
            for="Email"
          >
            Email
          </label>
          <div
            class="col-sm-9"
          >
            <input
              class="dropdown"
              id="Email"
              placeholder="Email"
              type="email"
              value=""
            />
          </div>
        </div>
        <div
          class="form-group row"
        >
          <label
            class="col-sm-3 col-form-label label"
            for="Password"
          >
            Password
          </label>
          <div
            class="col-sm-9"
          >
            <input
              class="dropdown"
              id="Password"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
        </div>
        <div
          class="container text-center"
        >
          <button
            class="undefined btn"
          >
            Submit
          </button>
        </div>
        <div
          class="mt-3"
        >
          New customer?
          <a
            class="link"
            href="/signup"
            id="register"
          >
            Sign up here!
          </a>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
